qiskit/synthesis/arithmetic/adders/vrg_modular_adder.py

```diff
@@ -0,0 +1,79 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2025.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compute modular sum of two qubit registers without any ancillary qubits."""
+
+from __future__ import annotations
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit import QuantumRegister
+from qiskit.circuit.library.standard_gates.x import XGate
+from qiskit.circuit.library.generalized_gates.mcmt import MCMTGate
+
+
+def adder_modular_v17(num_qubits: int) -> QuantumCircuit:
+    r"""
+    Construct a modular adder circuit with no ancillary qubits based on the Van Rentergem-style
+    adder in Fig. 15 of [1]. The implementation uses at most math:`16k - 15` CX gates.
+
+    Args:
+        num_qubits: The size of the register.
+
+    Returns:
+        The quantum circuit implementing the modular adder.
+
+    Raises:
+        ValueError: If ``num_qubits`` is less than 1.
+
+    **References:**
+
+    1. Gidney, Factoring with n+2 clean qubits and n-1 dirty qubits, 2017.
+    `arxiv:1706.07884` <https://arxiv.org/abs/1706.07884>`__
+
+    """
+
+    if num_qubits < 1:
+        raise ValueError("The number of qubits must be at least 1.")
+
+    qr_a = QuantumRegister(num_qubits, "a")
+    qr_b = QuantumRegister(num_qubits, "b")
+    qc = QuantumCircuit(qr_a, qr_b)
+
+    if num_qubits == 1:
+        qc.cx(qr_a[0], qr_b[0])
+        return qc
+
+    mcmt_a = MCMTGate(XGate(), 1, num_qubits - 1)
+    mcmt_b = MCMTGate(XGate(), 1, num_qubits)
+
+    qc.compose(mcmt_a, [qr_a[-1]] + qr_a[:-1], inplace=True)
+    qc.compose(mcmt_b, [qr_a[-1]] + qr_b[:], inplace=True)
+
+    # Ripple forward.
+    for i in range(num_qubits - 1):
+        qc.cx(qr_a[-1], qr_b[i])
+        qc.cx(qr_a[i], qr_a[-1])
+        qc.rccx(qr_b[i], qr_a[-1], qr_a[i])
+        qc.cx(qr_a[i], qr_a[-1])
```

Comment:
This looks a bit different from Fig. 15 (I have not done the maths, just looked at the picture). Is this some optimization to replace CX(a, b) CSWAP(b, a, c) by 3 CX + 1 RCCX gates?

Reply:
The first CX is part of the adder and for CSWAP I uesd the following definition:
```
CSWAP(a, b, c) = CX(c, b) CCX(a, b, c) CX(c, b)
```
and replaced CCX with RCCX because CSWAP gates appear in a compute-uncompute pattern:
<img width="293" height="411" alt="image" src="https://github.com/user-attachments/assets/581bf88f-3634-4d87-8dc3-cd755e79795b" />